cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/DefaultApplicationUpdateContext.java

```diff
@@ -0,0 +1,168 @@
+  }
+
+  @Override
+  public List<ApplicationConfigUpdateAction> getUpdateActions() {
+    return updateActions;
+  }
+
+  @Override
+  public <C extends Config> C getConfig(Type configType) {
+    // Given configtype has to be derived from Config class.
+    Preconditions.checkArgument(Config.class.isAssignableFrom(TypeToken.of(configType).getRawType()),
+                                "Application config type " + configType + " not supported. " +
+                                "Type must extend Config and cannot be parameterized.");
+    if (configString.isEmpty()) {
+      try {
+        return ((Class<C>) configType).newInstance();
+      } catch (Exception e) {
+        throw new IllegalArgumentException("Issue in creating config class of type " + configType.getTypeName(), e);
+      }
+    } else {
+      try {
+        return GSON.fromJson(configString, configType);
+      } catch (JsonSyntaxException e) {
+        throw new IllegalArgumentException("Invalid JSON application configuration was provided. Please check the"
+                                           + " syntax.", e);
+      }
+    }
+  }
+
+  @Override
+  public String getConfigAsString() {
+    return configString;
+  }
+
+  @Override
+  public List<ArtifactId> getPluginArtifacts(String pluginType, String pluginName, ArtifactScope pluginScope,
+                                             @Nullable ArtifactVersionRange pluginRange, int limit) throws Exception {
+    List<ArtifactId> pluginArtifacts = new ArrayList<>();
+    NamespaceId pluginArtifactNamespace = ArtifactScope.SYSTEM.equals(pluginScope) ? NamespaceId.SYSTEM : namespaceId;
+
+    Predicate<io.cdap.cdap.proto.id.ArtifactId> predicate = input -> {
+      // Check if it is from the scoped namespace and should check if plugin is in given range if provided.
+      return (pluginArtifactNamespace.equals(input.getParent()) &&
+             (pluginRange == null || pluginRange.versionIsInRange(new ArtifactVersion(input.getVersion()))));
+    };
+
+    try {
+      // TODO: Pass ArtifactSortOrder as argument for better flexibility.
+      Map<ArtifactDescriptor, PluginClass> plugins =
+        artifactRepository.getPlugins(pluginArtifactNamespace,
+                                      Artifact.from(Namespace.fromEntityId(namespaceId), applicationArtifactId),
+                                      pluginType, pluginName, predicate, limit, ArtifactSortOrder.ASC);
+      for (Map.Entry<ArtifactDescriptor, PluginClass> pluginsEntry : plugins.entrySet()) {
+        ArtifactId plugin = pluginsEntry.getKey().getArtifactId();
+        // Only consider non-SNAPSHOT plugins for upgrade.
+        // TODO: Consider making this check optional. Helpful for integration tests.
+        if (!plugin.getVersion().isSnapshot()) {
+          pluginArtifacts.add(plugin);
+        }
+      }
+    } catch (PluginNotExistsException e) {
+      LOG.debug("No plugin found for plugin {} of type {} in scope {} for app {}",
```

Comment:
This is because we are only looking for plugins in SYSTEM scope (even though current plugin is in USER scope). So there can be a case that the scope we are looking for does not have this plugin at all.

Reply:
If it is commonly expected to happen, log with trace instead of debug.